src/sagemaker/image.py

```diff
@@ -0,0 +1,509 @@
+        # first.
+        for channel in input_data_config:
+            uri = channel['DataSource']['S3DataSource']['S3Uri']
+            parsed_uri = urlparse(uri)
+            key = parsed_uri.path.lstrip('/')
+
+            channel_name = channel['ChannelName']
+            channel_dir = os.path.join(data_dir, channel_name)
+            os.mkdir(channel_dir)
+
+            if uri.lower().startswith("s3://"):
+                self._download_folder(bucket_name, key, channel_dir)
+            else:
+                volumes.append(Volume(uri, channel=channel_name))
+
+        # Create the configuration files for each container that we will create
+        # Each container will map the additional local volumes (if any).
+        for host in self.hosts:
+            _create_config_file_directories(self.container_root, host)
+            self.write_config_files(host, hyperparameters, input_data_config)
+            shutil.copytree(data_dir, os.path.join(self.container_root, host, 'input', 'data'))
+
+        compose_data = self._generate_compose_file('train', additional_volumes=volumes)
+        compose_command = self._compose()
+        _stream_output(compose_command)
+        _cleanup()
+
+        s3_model_artifacts = self.retrieve_model_artifacts(compose_data)
+
+        # free up the training data directory as it may contain
+        # lots of data downloaded from S3. This doesn't delete any local
+        # data that was just mounted to the container.
+        shutil.rmtree(data_dir)
+        # Also free the container config files.
+        for host in self.hosts:
+            shutil.rmtree(os.path.join(self.container_root, host))
+        return s3_model_artifacts
+
+    def serve(self, primary_container):
+        """Host a local endpoint using docker-compose.
+        Args:
+            primary_container (dict): dictionary containing the container runtime settings
+                for serving. Expected keys:
+                - 'ModelDataUrl' pointing to a local file
+                - 'Environment' a dictionary of environment variables to be passed to the hosting container.
+
+        """
+        logger.info("serving")
+
+        self.container_root = self._create_tmp_folder()
+        logger.info('creating hosting dir in {}'.format(self.container_root))
+
+        model_dir = primary_container['ModelDataUrl']
+        if not model_dir.lower().startswith("s3://"):
+            for h in self.hosts:
+                host_dir = os.path.join(self.container_root, h)
+                os.makedirs(host_dir)
+                shutil.copytree(model_dir, os.path.join(self.container_root, h, 'model'))
+
+        env_vars = ['{}={}'.format(k, v) for k, v in primary_container['Environment'].items()]
+
+        self._generate_compose_file('serve', additional_env_vars=env_vars)
+        compose_command = self._compose()
+        self.container = Container(self.container_root, compose_command)
+        self.container.up()
+
+    def stop_serving(self):
```

Comment:
When this refactored into its own SageMakerHostingContainer, consider making this be a context manager.